src/Transform.jl

```diff
@@ -0,0 +1,374 @@
+        vp[s][ids.kappa[2, :]] = 1.0 - vp[s][ids.kappa[1, :]]
+
+         # Positivity constraints
+        vp[s][ids.gamma] = exp(vp_free[s][ids_free.gamma])
+        vp[s][ids.zeta] = exp(vp_free[s][ids_free.zeta])
+    end
+end
+
+
+function free_unconstrain_sensitive_float(sf::SensitiveFloat, mp::ModelParams)
+    # Given a sensitive float with derivatives with respect to all the
+    # constrained parameters, calculate derivatives with respect to
+    # the unconstrained parameters.
+    #
+    # Note that all the other functions in ElboDeriv calculated derivatives with
+    # respect to the unconstrained parameterization.
+
+    # Require that the input have all derivatives defined.
+    @assert size(sf.d) == (ids.size, mp.S)
+
+    sf_free = zero_sensitive_float(collect(1:mp.S), CelesteTypes.all_params_free)
+    sf_free.v = sf.v
+
+    # Currently the param_index is only really used within ElboDeriv.  By the
+    # time the data hits the optimizer, we assume everything has a derivative. 
+    sf_free.param_index = all_params_free
+
+    for s in 1:mp.S
+        # Variables that are unaffected by constraints:
+        for id_string in rect_unchanged_ids
+            id_symbol = convert(Symbol, id_string)
+
+            # Flatten the indices for matrix indexing
+            id_free_indices = reduce(vcat, ids_free.(id_symbol))
+            id_indices = reduce(vcat, ids.(id_symbol))
+
+            sf_free.d[id_free_indices, s] = sf.d[id_indices, s]
+        end
+
+        # TODO: write in general form.  Note that the old "chi" is now chi[2].
+        # Simplicial constriants.
+        this_chi = mp.vp[s][ids.chi[2]]
+        sf_free.d[ids_free.chi[1], s] =
+            (sf.d[ids.chi[2], s] - sf.d[ids.chi[1], s]) * this_chi * (1.0 - this_chi)
+
+        this_kappa = collect(mp.vp[s][ids.kappa[1, :]])
+        sf_free.d[collect(ids_free.kappa[1, :]), s] =
+            (sf.d[collect(ids.kappa[1, :]), s] - sf.d[collect(ids.kappa[2, :]), s]) .*
+            this_kappa .* (1.0 - this_kappa)
+
+        # Positivity constraints.
+        sf_free.d[ids_free.gamma, s] = sf.d[ids.gamma, s] .* mp.vp[s][ids.gamma]
+        sf_free.d[ids_free.zeta, s] = sf.d[ids.zeta, s] .* mp.vp[s][ids.zeta]
+    end
+
+    sf_free
+end
+
+rect_transform = DataTransform(rect_to_vp!, vp_to_rect!,
+                               vector_to_free_vp!, free_vp_to_vector, 
+                               rect_unconstrain_sensitive_float,
+                               ids_free.size)
+
+free_transform = DataTransform(free_to_vp!, vp_to_free!,
+                               vector_to_free_vp!, free_vp_to_vector,
+                               free_unconstrain_sensitive_float, ids_free.size)
+
+identity_transform = DataTransform(unchanged_vp!, unchanged_vp!,
+                                   unchanged_vector_to_vp!, unchanged_vp_to_vector,
+                                   unchanged_sensitive_float, ids.size)
+
```

Comment:
Nice...looks like this approach makes alternate parametrizations straightforward